packages/connections/src/stores/connections-store.ts

```diff
@@ -0,0 +1,275 @@
+      return {
+        ...state,
+        connectionAttempt: null,
+        connectionErrorMessage: 'Connection attempt canceled.',
+      };
+    case 'connection-attempt-succeeded':
+      return {
+        ...state,
+        connectionAttempt: null,
+        isConnected: true,
+      };
+    case 'connection-attempt-errored':
+      return {
+        ...state,
+        connectionAttempt: null,
+        connectionErrorMessage: action.connectionErrorMessage,
+      };
+    case 'set-active-connection':
+      return {
+        ...state,
+        activeConnectionId: action.connectionId,
+        activeConnectionInfo: action.connectionInfo,
+      };
+    case 'new-connection':
+      return {
+        ...state,
+        activeConnectionId: action.connectionInfo.id,
+        activeConnectionInfo: action.connectionInfo,
+      };
+    case 'set-connections':
+      return {
+        ...state,
+        connections: action.connections,
+      };
+    default:
+      return state;
+  }
+}
+
+export function useConnections(
+  onConnected: (
+    connectionInfo: ConnectionInfo,
+    dataService: DataService
+  ) => Promise<void>
+): [
+  State,
+  {
+    onCancelConnectionAttempt(): void;
+    onConnect(connectionInfo: ConnectionInfo): Promise<void>;
+    updateActiveConnection(newConnectionId?: string | undefined): void;
+  }
+] {
+  const [state, dispatch] = useReducer(
+    connectionsReducer,
+    defaultConnectionsState()
+  );
+  const { isConnected, connectionAttempt, connections } = state;
+
+  const connectingConnectionAttempt = useRef<ConnectionAttempt>();
+  const connectedConnectionInfo = useRef<ConnectionInfo>();
+  const connectedDataService = useRef<DataService>();
+
+  useEffect(() => {
+    if (
+      isConnected &&
+      connectedConnectionInfo.current &&
+      connectedDataService.current
+    ) {
+      // After connecting and the UI is updated we notify the rest of Compass.
+      void onConnected(
+        connectedConnectionInfo.current,
+        connectedDataService.current
+      );
+    }
+  }, [isConnected, onConnected]);
+
+  useEffect(() => {
+    async function loadConnections() {
+      try {
+        const connectionStorage = new ConnectionStorage();
+        const loadedConnections = await connectionStorage.loadAll();
+
+        dispatch({
+          type: 'set-connections',
+          connections: loadedConnections,
+        });
+      } catch (error) {
+        debug('error loading connections', error);
+      }
+    }
+
+    // Load connections after first render.
+    void loadConnections();
+
+    return () => {
+      // When unmounting, clean up any current connection attempts that have
+      // not resolved.
+      if (
+        connectingConnectionAttempt.current &&
+        !connectingConnectionAttempt.current.isClosed()
+      ) {
+        connectingConnectionAttempt.current.cancelConnectionAttempt();
+      }
+    };
+  }, []);
+
+  return [
+    state,
+    {
+      onCancelConnectionAttempt() {
```

Comment:
It could be hard to read / edit these inside an object that is inside an array, what about defining these functions outside: 

```
fn onCancelConnectionAttempt() {}

return [state, {onCancelConnectionAttempt, ...}]
```